text-serializer-gson/src/main/java/net/kyori/adventure/text/serializer/gson/ShowItemSerializer.java

```diff
@@ -23,71 +23,82 @@
+  public static final TypeAdapter<HoverEvent.ShowItem> INSTANCE = new ShowItemSerializer().nullSafe();
+
   static final String ID = "id";
   static final String COUNT = "count";
   static final String TAG = "tag";
 
-  ShowItemSerializer() {
+  private ShowItemSerializer() {
   }
 
   @Override
-  public HoverEvent.ShowItem deserialize(final JsonElement json, final Type typeOfT, final JsonDeserializationContext context) throws JsonParseException {
-    final JsonObject object = json.getAsJsonObject();
-
-    if(!object.has(ID)) {
-      throw new JsonParseException("Not sure how to deserialize show_item hover event");
-    }
-
-    final Key id = context.deserialize(object.getAsJsonPrimitive(ID), Key.class);
+  public HoverEvent.ShowItem read(final JsonReader in) throws IOException {
+    in.beginObject();
 
+    Key key = null;
     int count = 1;
-    if(object.has(COUNT)) {
-      count = object.get(COUNT).getAsInt();
-    }
+    @Nullable BinaryTagHolder nbt = null;
 
-    BinaryTagHolder nbt = null;
-    if(object.has(TAG)) {
-      final JsonElement tag = object.get(TAG);
-      if(tag.isJsonPrimitive()) {
-        nbt = BinaryTagHolder.of(tag.getAsString());
-      } else if(!tag.isJsonNull()) {
-        throw new JsonParseException("Expected " + TAG + " to be a string");
+    while(in.hasNext()) {
+      final String fieldName = in.nextName();
+      if(fieldName.equals(ID)) {
+        key = KeySerializer.INSTANCE.read(in);
+      } else if(fieldName.equals(COUNT)) {
+        count = in.nextInt();
+      } else if(fieldName.equals(TAG)) {
+        final JsonToken token = in.peek();
+        if(token == JsonToken.STRING || token == JsonToken.NUMBER) {
+          nbt = BinaryTagHolder.of(in.nextString());
+        } else if(token == JsonToken.BOOLEAN) {
+          nbt = BinaryTagHolder.of(String.valueOf(in.nextBoolean()));
+        } else if(token == JsonToken.NULL) {
+          in.nextNull();
+        } else {
+          throw new JsonParseException("Expected " + TAG + " to be a string");
+        }
+      } else {
+        in.skipValue();
       }
     }
 
-    return HoverEvent.ShowItem.of(id, count, nbt);
+    if(key == null) {
+      throw new JsonParseException("Not sure how to deserialize show_item hover event");
+    }
+    in.endObject();
+
+    return HoverEvent.ShowItem.of(key, count, nbt);
   }
 
   @Override
-  public JsonElement serialize(final HoverEvent.ShowItem src, final Type typeOfSrc, final JsonSerializationContext context) {
-    final JsonObject json = new JsonObject();
+  public void write(final JsonWriter out, final HoverEvent.ShowItem value) throws IOException {
+    out.beginObject();
 
-    json.add(ID, context.serialize(src.item()));
+    out.name(ID);
+    KeySerializer.INSTANCE.write(out, value.item());
 
-    final int count = src.count();
+    final int count = value.count();
     if(count != 1) {
-      json.addProperty(COUNT, count);
+      out.name(COUNT).value(count);
```

Comment:
I think I'd prefer to split this (applies to all cases) to be consistent with cases like above where we delegate to another serialiser:
```suggestion
      out.name(COUNT);
      out.value(count);
```

Reply:
https://github.com/KyoriPowered/adventure/pull/373/commits/4bc69716006f639c63a19847be2b895bd5065ab8